cmd/influx_inspect/deletetsm/deletetsm.go

```diff
@@ -68,77 +70,100 @@ func (cmd *Command) Run(args ...string) (err error) {
+			return 0, err
+		}
 
-	// Iterate over the input blocks.
-	itr := r.BlockIterator()
-	for itr.Next() {
-		// Read key & time range.
-		key, minTime, maxTime, _, _, block, err := itr.Read()
+		// Create TSMWriter to temporary location.
+		output, err := os.Create(outputPath)
 		if err != nil {
-			return err
+			return 0, err
 		}
 
-		// Skip block if this is the measurement and time range we are deleting.
-		series, _ := tsm1.SeriesAndFieldFromCompositeKey(key)
-		measurement, tags := models.ParseKey(series)
-		if string(measurement) == cmd.measurement || (cmd.sanitize && !models.ValidKeyTokens(measurement, tags)) {
-			log.Printf("deleting block: %s (%s-%s) sz=%d",
-				key,
-				time.Unix(0, minTime).UTC().Format(time.RFC3339Nano),
-				time.Unix(0, maxTime).UTC().Format(time.RFC3339Nano),
-				len(block),
-			)
-			continue
+		w, err := tsm1.NewTSMWriter(output)
+		if err != nil {
+			// close the output file on error creating the TSMWriter
+			_ = output.Close()
+			return 0, fmt.Errorf("unable to write %s: %w", outputPath, err)
 		}
 
-		if err := w.WriteBlock(key, minTime, maxTime, block); err != nil {
-			return err
+		// This will close the output file
+		defer errors2.Capture(&fRetErr, w.Close)()
+
+		// Iterate over the input blocks.
+		itr := r.BlockIterator()
+		blockWritten := false
+		for itr.Next() {
+			// Read key & time range.
+			key, minTime, maxTime, _, _, block, err := itr.Read()
+			if err != nil {
+				return 0, err
+			}
+
+			// Skip block if this is the measurement and time range we are deleting.
+			series, _ := tsm1.SeriesAndFieldFromCompositeKey(key)
+			measurement, tags := models.ParseKey(series)
+			if string(measurement) == cmd.measurement || (cmd.sanitize && !models.ValidKeyTokens(measurement, tags)) {
+				log.Printf("deleting block: %s (%s-%s) sz=%d",
+					key,
+					time.Unix(0, minTime).UTC().Format(time.RFC3339Nano),
+					time.Unix(0, maxTime).UTC().Format(time.RFC3339Nano),
+					len(block),
+				)
+				continue
+			}
+
+			if err := w.WriteBlock(key, minTime, maxTime, block); err != nil {
+				return 0, err
+			}
+			blockWritten = true
 		}
-	}
 
-	// Write index & close.
-	if err := w.WriteIndex(); err != nil {
-		return err
-	} else if err := w.Close(); err != nil {
+		// Write index & close.
+		// It is okay to have no index values if no block was written
+		if err := w.WriteIndex(); err != nil && !(blockWritten || errors.Is(err, tsm1.ErrNoValues)) {
+			return 0, err
+		} else {
+			return w.Size(), nil
+		}
+	}()
+	if err != nil {
 		return err
 	}
 
-	// Replace original file with new file.
-	return os.Rename(outputPath, path)
+	if size > 0 {
+		// Replace original file with new file.
+		return os.Rename(outputPath, path)
+	} else if err = os.RemoveAll(path); err != nil {
+		return fmt.Errorf("cannot remove %s: %w", path, err)
+	} else if err = os.RemoveAll(outputPath); err != nil {
+		return fmt.Errorf("cannot remove temporary file %s: %w", outputPath, err)
```

Comment:
If we can't remove `path`, then we won't try to remove `outputPath`? Should we be trying to delete `outputPath` even if we fail to remove `path`?